harness/determined/pytorch/samplers.py

```diff
@@ -0,0 +1,361 @@
+
+- Even if you are going to ultimate return an IterableDataset, it is best to use PyTorch's Sampler
+  class as the basis for choosing the order of which records to read.  Operations on Samplers are
+  quick and cheap, while operations on data afterwards are expensive.  For more details, see the
+  discussion at TODO:ref_yogadl_readme.  If you don't have a custom sampler, start with a simple
+  one:
+
+  .. code::python
+
+     sampler = torch.utils.data.SequentialSampler(my_dataset)
+
+- **Shuffle first**: Always use a reproducible shuffle when you shuffle. Determined provides a
+  ReproducibleShuffleSampler for this purpose.  Determined also provides a
+  ReproducibleShuffleBatchSampler class, but you should prefer to shuffle at the record level
+  whenever possible, to achieve the highest-quality shuffle.
+
+- **Repeat when training**: In Determined, you always repeat your training dataset and you
+  never repeat your validation datasets.  Determined provides a RepeatSampler and a
+  RepeatBatchSampler to wrap your sampler or batch_sampler.  For your training dataset, make sure
+  that you always repeat AFTER you shuffle, otherwise your shuffle will hang.
+
+- **Always shard, and not before a repeat**: Use Determined's DistributedSampler or
+  DistributedBatchSampler to provide a unique shard of data to each worker based on your sampler or
+  batch_sampler.  It is best to always shard your data, and even when you are not doing distributed
+  training, because in non-distributed-training settings, the sharding is nearly zero-cost, and it
+  makes distributed training seamless if you ever want to use it in the future.
+
+  It is generally important to shard after you repeat, unless you can guarantee that each shard
+  of the dataset will have the same length.  Otherwise, differences between the epoch boundaries for
+  each worker can grow over time, especially on small datasets.  If you shard after you repeat, you
+  can change the number of workers arbitrarily without issue.
+
+- **Skip when training, and always last**: In Determined, training datasets should always be able to
+  start from an arbitrary point in the dataset.  This allows for advanced hyperparameter searches
+  and responsive preemption for training on spot instances in the cloud.  The easiest way to do
+  this, which is also very efficient, is to apply a skip to the sampler.
+
+  Determined provides a SkipBatchSampler that you can apply to your batch_sampler for this purpose.
+  There is also a SkipSampler that you can apply to your sampler, but you should prefer to skip on
+  batches unless you are confident that your dataset always yields identical size batches, where the
+  number of records to skip can be reliably calculatd from the number of batches already trained.
+
+  Always skip AFTER your repeat, so that the skip only happens once, and not on every dataset.
+
+  Always skip AFTER your shuffle, to preserve the reproducibility of the shuffle.
+
```

Comment:
can we add a quick bullet point that reiterates above of the order i need. This way when using as a reference its like a recipe.

Reply:
Um... I'm a bit hesitant to do that because I feel like if I say:

> Remember: Shuffle-repeat-shard-batch-skip and you're good

then that simplifies things to the point of being confusing, like "why would I have to shuffle and repeat my validation dataset?"

I tried really hard to let the bolded parts of the bullets be a simple checklist:

* **Shuffle first**: ...
* **Repeat when training**: ...
* **Always shard, and not before a repeat**: ...
* **Skip when training, and always last**: ...

I'm not sure that it's very safe to simplify it any simpler than that...

Reply:
Okay. Lets leave it then as is.